File: src/breakpoints.rs
use crate::context::*;
use crate::debug_adapter_comms;

use json::{JsonValue, object};

pub fn process_breakpoints(ctx: &mut Context, breakpoints: JsonValue) {
    // If we have no breakpoints, don't do anything
    if breakpoints["breakpoints"].is_empty() {
        return;
    }
    let raw_breakpoints = breakpoints["breakpoints"].to_string();
    // Split passed value along spaces
    let split = raw_breakpoints.split_whitespace();
    for val in split {
        // Get the filepath and line number
        let bar_index = val.chars().position(|c| c == '|').unwrap();
        let line_no = &val[0..bar_index].parse::<u64>().unwrap();
        let filepath = &val[bar_index + 1..];
        debug!("Filepath: {}, line_no: {}", filepath, line_no);
        // If we already have breakpoints in this filename
        if ctx.breakpoint_data.contains_key(filepath) {
            let mut new_vec = ctx.breakpoint_data.get(filepath).unwrap().clone();
            new_vec.push(*line_no);
            ctx.breakpoint_data.insert(filepath.to_string(), new_vec);
        }
        // Otherwise, create a new entry
        else {
            let mut new_vec = vec![];
            new_vec.push(*line_no);
            ctx.breakpoint_data.insert(filepath.to_string(), new_vec);
        }
    }
}

// Handles the "initialized" event.
pub fn handle_initialized_event(_msg: json::JsonValue, ctx: &mut Context) {
    // This is where we set the breakpoints
    let mut requests : Vec<json::JsonValue> = vec![];
    // Loop over the various source files we were sent
    for (source, lines) in &ctx.breakpoint_data {
        let mut breakpoints = json::JsonValue::new_array();
        // Ensure we get all the lines in each file
        for line in lines {
            breakpoints.push(object! {
                "line": *line,
            }).expect("Couldn't add breakpoint to list");
        }
        // Construct the actual request's arguments
        let mut break_args = object! {
            "source": {
                "path": source.to_string(),
            },
        };
        break_args["breakpoints"] = breakpoints;
        requests.push(break_args);
    }
    // Send all the breakpoint requests, one after another
    for req in requests {
        debug_adapter_comms::do_request("setBreakpoints", req, ctx);
    }
  
    // Now, send the configurationDone request.
    debug_adapter_comms::do_request("configurationDone", object! {}, ctx);
}

File: src/breakpoints.py
import debug_session
import general
import logging
import os

breakpoint_data = {}  # filepath: line_nos


def process_breakpoints():
    global breakpoint_data
    # First ensure we have breakpoints to retrieve
    if "kak_opt_dap_breakpoints_info" not in os.environ:
        return
    # Retrieve breakpoints
    breakpoints_info = os.environ["kak_opt_dap_breakpoints_info"]
    logging.debug(f"Breakpoint string: {breakpoints_info}")
    split_breaks = breakpoints_info.split()
    # Parse breakpoints
    for val in split_breaks:
        break_vals = val.split("|")
        line_no = break_vals[0]
        filepath = break_vals[1]
        logging.debug(f"Breakpoint filepath: {filepath}, line no: {line_no}")
        # If we already have breakpoints under this filepath, just add another
        if filepath in breakpoint_data.keys():
            breakpoint_data[filepath].append(line_no)
        # Otherwise, add a new entry for this filepath
        else:
            breakpoint_data[filepath] = [line_no]


def handle_initialized_event(msg):
    # We need to send breakpoint requests after the initialized event
    # For background: https://github.com/microsoft/vscode/issues/4902
    requests = []
    for source in breakpoint_data:
        breakpoints = []
        lines = breakpoint_data[source]
        for line in lines:
            breakpoints.append({"line": int(line)})
        break_args = {"source": {"path": source}, "breakpoints": breakpoints}
        requests.append(break_args)
    for req in requests:
        debug_session.debug_adapter.write_request(
            "setBreakpoints", req, lambda *args: None
        )

    # Now send the configurationDone request.
    # We only do this if the adapter has advertised the "configurationDone"
    # capability.
    if (
        general.capabilities is not None
        and "supportsConfigurationDoneRequest" in general.capabilities
        and general.capabilities["supportsConfigurationDoneRequest"]
    ):
        debug_session.debug_adapter.write_request(
            "configurationDone", {}, lambda *args: None
        )
